File: aws/update_sensor.py
import json
import os
import pymysql


def lambda_handler(event, context):
    # Informations de connexion à la base de données
    host = os.environ['HOST']
    username = os.environ['USERNAME']
    password = os.environ['PASSWORD']
    db_name = os.environ['DB_NAME']
    table_capt = os.environ['TABLE_CAPT']
    status_code = 200
    body = "success"

    sensor_id = event['params']['name']
    body = event['body']

    # Connexion à la base de données
    conn = pymysql.connect(db=db_name, user=username, passwd=password, host=host, port=3306)

    try:
        with conn.cursor() as cursor:
            update_sensor = f"UPDATE {table_capt} SET threshold = '{body['threshold']}' timer = '{body['timer']}' " \
                            f"WHERE sensor_id = '{sensor_id}';"
            cursor.execute(update_sensor)
        conn.commit()

        response = {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',  # Allow requests from any origin
                'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',  # Allow the specified HTTP methods
                'Access-Control-Allow-Headers': 'Content-Type'  # Allow the specified headers
            },
            'body': json.dumps({
                'message': 'Données enregistrées'
            })
        }
        return response

    except pymysql.MySQLError as e:
        logger.error("ERROR: Unexpected error: Could not connect to MySQL instance.")
        logger.error(e)
        sys.exit()

    finally:
        conn.close()

File: backend/lambdas/update-sensor.ts
import {
  APIGatewayProxyEvent,
  APIGatewayProxyResult,
  Context,
} from "aws-lambda";
import AWS from "aws-sdk";
import mysql from "mysql2/promise";

export const handler = async (
  event: APIGatewayProxyEvent,
  context: Context
): Promise<APIGatewayProxyResult> => {
  const regexGeneric = /^[a-zA-Z0-9\s,.'\-+éàè?î!âôêû:\/ù]{1,}$/;
  const regexNumber = /^[0-9]*$/;

  const badRequest: any = { message: "Requête invalide." };
  badRequest.status = 400;

  let body: any;
  let statusCode = 200;
  const headers: any = {
    "Content-Type": "application/json",
    "Access-Control-Allow-Origin": "*", // Allow requests from any origin, you can restrict it to specific origins if needed
    "Access-Control-Allow-Methods": "PUT", // Specify the allowed HTTP methods
    "Access-Control-Allow-Headers": "Content-Type",
  }; // Specify the allowed headers
  let connection: any;

  try {
    const sensorId = event.pathParameters.sensorId;

    // validation des données entrantes
    if (!sensorId || !regexNumber.test(sensorId)) {
      throw badRequest;
    }

    const data = JSON.parse(event.body);
    console.log("data", data);

    if (
      !data.name ||
      !regexGeneric.test(data.name) ||
      !data.threshold ||
      !regexNumber.test(data.threshold) ||
      !data.timer ||
      !regexNumber.test(data.timer)
    ) {
      throw badRequest;
    }

    const connectionConfig = {
      host: process.env.HOST,
      user: process.env.USER,
      password: process.env.PASSWORD,
      database: process.env.DATABASE,
      socketPath: process.env.SOCKET_PATH,
    };
    const table1 = process.env.TABLE_1;

    connection = await mysql.createConnection(connectionConfig);

    let stmt = `UPDATE ${table1} SET name=?, threshold=?, timer=? WHERE id=?`;
    const [result] = await connection.query(stmt, [
      data.name,
      data.threshold,
      data.timer,
      sensorId,
    ]);

    body = JSON.stringify(result);
  } catch (error: any) {
    statusCode = error.status ?? 500;
    body = JSON.stringify({
      message: error.message ?? "Un problème est survenu.",
    });
  } finally {
    if (connection) {
      await connection.end();
    }
  }
  return {
    statusCode,
    body,
    headers,
  };
};
